### PR TITLE
Add a jquery rating plugin

### PR DESCRIPTION
**Pull request by danjee**
The github page:
https://github.com/dreyescat/bootstrap-rating
Last commit was 3 months ago, so it seems the project is alive

And the demo page: http://dreyescat.github.io/bootstrap-rating/

### CODE REVIEW

File: bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import org.slf4j.LoggerFactory;
+
+import de.agilecoders.wicket.webjars.request.resource.WebjarsCssResourceReference;
+
+/**
+ * Bootstrap Rating is a jQuery plugin that creates a rating control that uses Bootstrap glyphicons for rating symbols. 
+ * http://dreyescat.github.io/bootstrap-rating
+ * @author daniel.jipa
+ */
+public class Rating<T> extends HiddenField<T>{
```
Comment (martin-g) [inline bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java]: I'd prefer to rename it to `RatingField`. Just `Rating` sounds like a POJO/entity to me.
WDYT?

File: bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	private static final long serialVersionUID = 3927736784638450199L;
+	
+	private static final Logger LOG = LoggerFactory.getLogger(Rating.class);
+	
+	private static final CssResourceReference RATING_CSS = new CssResourceReference(Rating.class,
+			"css/bootstrap-rating.css");
+	private static final JavaScriptResourceReference RATING_JS = new JavaScriptResourceReference(Rating.class,
+			"js/bootstrap-rating.min.js") {
+
+		private static final long serialVersionUID = -7280204736754448532L;
```
Comment (martin-g) [inline bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java]: Duplicate

File: bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ */
+public class Rating<T> extends HiddenField<T>{
+
+	private static final long serialVersionUID = 3927736784638450199L;
+	
+	private static final Logger LOG = LoggerFactory.getLogger(Rating.class);
+	
+	private static final CssResourceReference RATING_CSS = new CssResourceReference(Rating.class,
+			"css/bootstrap-rating.css");
+	private static final JavaScriptResourceReference RATING_JS = new JavaScriptResourceReference(Rating.class,
```
Comment (martin-g) [inline bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java]: Use `JQueryPluginResourceReference` instead to skip the dependency below.

File: bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java (hunk 1 above)
Comment (danjee) [inline bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java]: It's OK for me. Should I rename the config class as well ?

File: bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        this(id, null, config);
+    }
+
+    public Rating(String id, final IModel<T> model) {
+        this(id, model, new RatingConfig());
+    }
+
+    public Rating(String id, final IModel<T> model, RatingConfig config) {
+        super(id, model);
+        this.config = config;
```
Comment (martin-g) [inline bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java]: `Args.notNull(config, "config")`

File: bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    
+    private void registerChangeListener() {
+        add(new AjaxEventBehavior("change"){
+
+        	private static final long serialVersionUID = -8135111313923624591L;
+
+			@Override
+            protected void onEvent(AjaxRequestTarget target) {
+				if (wantChangeListener()){
+					LOG.info("change rate triggered");
```
Comment (martin-g) [inline bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java]: This is rather debug. not info.

File: bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    
+    private void registerChangeListener() {
+        add(new AjaxEventBehavior("change"){
+
+        	private static final long serialVersionUID = -8135111313923624591L;
+
+			@Override
+            protected void onEvent(AjaxRequestTarget target) {
+				if (wantChangeListener()){
```
Comment (martin-g) [inline bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java]: IMO `wantChangeListener()` should be around `registerChangeListener()` in `#onInitialize()`. No need to bind an Ajax behavior at all.

File: bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+					LOG.info("change rate triggered");
+	                onChange(target);
+				}
+            }
+        });
+    }
+    
+	protected boolean wantChangeListener() {
```
Comment (martin-g) [inline bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java]: Rename to `wantAjaxNotification()` ?!

File: bootstrap-samples/src/main/java/de/agilecoders/wicket/samples/pages/ExtensionsPage.java [hunk 8]
```diff
@@ -357,6 +384,8 @@ public void renderHead(IHeaderResponse response) {
         super.renderHead(response);
 
         response.render(CssHeaderItem.forReference(OpenWebIconsCssReference.instance()));
+		response.render(CssHeaderItem.forReference(new WebjarsCssResourceReference("font-awesome/current/css/font-awesome.css")));
```
Comment (martin-g) [inline bootstrap-samples/src/main/java/de/agilecoders/wicket/samples/pages/ExtensionsPage.java]: You could use `FontAwesomeCssResourceReference` (or something similar). We have such resref in -extensions already.

**Comment by martin-g** [comment]
Awesome!
Many thanks!

I've added some minor suggestions.

File: bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java (hunk 1 above)
Comment (martin-g) [inline bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/Rating.java]: RatingConfig is OK!

File: bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/RatingCssResourceReference.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.apache.wicket.request.resource.PackageResourceReference;
+
+import de.agilecoders.wicket.extensions.slider.res.BootstrapSliderCssResourceReference;
+
+public class RatingCssResourceReference extends PackageResourceReference {
+
+	private static final long serialVersionUID = 1L;
+
+	private static final BootstrapSliderCssResourceReference INSTANCE = new BootstrapSliderCssResourceReference();
```
Comment (danjee) [inline bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/RatingCssResourceReference.java]: I have an issue here due to bad copy-paste. I will fix it and make the PR again

File: bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/RatingCssResourceReference.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
+package de.agilecoders.wicket.extensions.markup.html.bootstrap.form.rating;
+
+import org.apache.wicket.request.resource.PackageResourceReference;
+
+
+public class RatingCssResourceReference extends PackageResourceReference {
```
Comment (martin-g) [inline bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/rating/RatingCssResourceReference.java]: This should be `CssResourceReference`.
I'll fix it!

**Comment by danjee** [comment]
Thanks!

**Comment by martin-g** [comment]
Why the `step` is `double` ?
At https://github.com/dreyescat/bootstrap-rating/blob/master/bootstrap-rating.js#L24 I see it is parsed as `int`

**Comment by martin-g** [comment]
I've made some minor improvements.
Please let me know if something is not correct!

Thanks again for the contribution!

**Pull request merged**